Feature/page numbers (#171)
j3-fortran/generics: J3-Papers/edits/misc.txt

new new clause introduced in papers 1-5 that are necessary to incorporate
the syntax and semantics for templates.

2. Misc edits for templates
===========================

Clause 3:  Terms and definitions
--------------------------------

[5:4+] Insert new glossary term:

  "3.7.2b
   [deferred argument association]
   association between an instantiation argument and a deferred
   argument (Sec. ???)"

[10:9+] Insert new glossary term

  "3.34.1
  [deferred constant]
  a deferred argument with the PARAMETER attribute"

[11:32+] Insert new glossary terms

  "3.46b
   [deferred argument]
   named entity that appears in a <deferred-arg-list> (R???) (Sec. ???)"

[17:18+] Insert new glossary term

  "3.85b
  [instantiation argument]
  entity that determines deferred argument association"

[32:16+] Insert new glossary term

  "3.151b
  [ultimately defined prior]
  an entity that is accessed by host association and whose definition
  appears before the scoping unit in which it is reference or is
  ultimately defined prior in the host scoping unit"

> {UTI - is this still used? TLC}

Reviewer: This one should go after the following two edits.

Author: Yup.  Fixing in sandbox.

       "Each program unit, module subprogram, templated subprogram,
       and internal subprogram shall have exactly one END statement."

[49:7] After "end-module-statement", insert ", end-template-statement,
       end-requirement-stmt" such that the sentence now reads:

        "The end-module-stmt, end-template-statement,
        end-requirement-stmt, end-submodule-stmt, and
        end-block-data-stmt statements are nonexecutable."

* 5.4.1.1 General

[51:13] 5.4.1.1 Geeneral

  Modify first sentence of paragraph 2 to read:

  "A type is either an intrinsic type, a derived type, or a deferred type."

{UTI - this sentence is very different than what is in the standard.}

[51:27+]  Add a new subclause

  "5.4.1.3+ Deferred type

  Deferred types are deferred arguments that are types in templates,
  templated procedures, or requirements.  Deferred types are described
  further in tt.xx. "

Clause 6: Lexical tokens and source form
----------------------------------------

[62:4] 6.2.6 Delimiters
   Change  "or /)"  to  "/), {, or }." such that the entire sentence reads:

  "A lexical token that is a delimiter is a (, ), /, [, ], {, }, (/, or /)."

> <<<<got to here>>>>>

Reviewer: ```suggestion
```

       <<or>> TYPEOF ( <data-ref> )
       <<or>> CLASSOF ( <data-ref> )"

> [68-31] Change C705 wording to be consintent with constraints just after it:

Reviewer: ```suggestion
[68:31] Change C705 wording to be consintent with constraints just after it:
```

different than the nigh identical constraint that follows, that
generics subgroup thought it should be updated.}

[68:34+] After C705, insert a new constraint:

   "C705+ (R703) CLASS(<deferred-type>) shall specify an extensible
    type (7.5.7)."

[69:1+] After C706 introduce a new constraint

   "C706+ (R703) TYPE(<deferred-type>) shall not specify an abstract type
   (7.5.7)."

[80:27+] In 7.5.2.1 Syntax of a derived-type definition, insert a new

>          consntraint on R728 afte C737:

Reviewer: ```suggestion
         constraint on R728 after C737:
```

           DEFERRED INTEGER, PARAMETER :: S(*)
           DEFERRED INTEGER, PARAMETER :: C(..) ! implied-rank
           INTEGER, PARAMETER :: B(S) = 1       ! implied-rank

>     REAL, PARAMETER, RANK(RANK(C)) :: D  ! implied rank

Reviewer: ```suggestion
           REAL, PARAMETER, RANK(RANK(C)) :: D  ! implied rank
```

Author: Ah thanks.   I know when I replaced <tab> with space in the document I saw one misformatted line go bast, but could not spot it when I went back.

    bounds equal to one; otherwise, it specifies that it is an
    assumed-shape array with all the lower bounds equal to one".

[128:36-37] In 8.6.1 Accessibility Statement, in constraint C873, after
            "a module", insert " or template" twice, such that the
            constraint reads:

  "C873 (R831) An <access-stmt> shall appear only in the
        <specification-part> of a module or template. Only one
        accessibility statement with an omitted <access-id-list> is
        permitted in the <specification-part> of a module or
        template."

> [129:1-2] In constraint, C874, replace "or namelist group"
>           namelist-group, requirement, or template" such that the

Reviewer: ```suggestion
[129:1-2] In constraint, C874, replace "or namelist group" with
          "namelist-group, requirement, or template" such that the
```

    subprogram is the null mapping.
   "

[139:7] In section 8.8 IMPORT Statement, in second sentence of p2,
        after "module procedure interface body", insert " or an interface body
        that appears in a deferred interface block."  The sentence will then read:

    "This is the default for an interface body that is not a module
     procedure interface body or an interface body that appears in a
     deferred interface block."

[139:14] In sentence 2 of p4, replace "or submodule" with "submodule,
         or an interface body that appears in a deferred interface
         block", such that it now reads:

    "This is the default for a derived-type definition, internal
     subprogram, module procedure interface body, module subprogram,
     submodule, or an interface body that appears in a deferred
     interface block."

Clause 10: Expressions and  assignment
--------------------------------------

[187:4] In, 10.1.12 Constant expression remove " or" from end of item
        (14), and " or" to end of item (15) and then introduce a new
        item for deferred constant.  The lines should now read:

  (14) an ac-do-variable within an array constructor where each

>        scalar-int-expr of the correspo nding ac-implied-do-control is

Reviewer: ```suggestion
       scalar-int-expr of the corresponding ac-implied-do-control is
```

       a constant expression,
  (15) a constant expression enclosed in parentheses, or
  (16) a deferred constant,

{This change is possibly not necessary.  Subgroup is not certain.}

* 10.2.1.2 Intrinsic assignment statement

{UTI Subgroup believes that the existing constraints on intrinsic
assignment in 10.2.1.2 "Intrinsic assigment statement" are sufficient
for the case of intrinsic assignment with deferred rank entities.  But
would appreciate further feedback on this issue.}

Clause 11: Execution control
----------------------------

[225:17+] In section 11.1.11.1 Purpose and form of the SELECT TYPE
          construct, insert entry in R1156 type-guard-stmt:

>   "<<or>> CLASS IS ( <derived-type-spec> ) [ <select-construct-name> ]"

Reviewer: ```suggestion
  "<<or>> CLASS IS ( <deferred-type> ) [ <select-construct-name> ]"
```

              <<or>> <templated-function-subprogram>
              <<or>> <templated-subroutine-subprogram>"

> [326:2] In section 14.2.2 The USE statement and use assocation,

Reviewer: ```suggestion
[327:2] In section 14.2.2 The USE statement and use assocation,
```

{UTI: Need a constraint that only deferred procedures can appear in a
deferred interface block.

[335:23+] Introduce new constraint on rule 1503 interface-stmt:

      "C1503b Each <subroutine-name> or <function-name> of an
       <interface-body> that appears in a deferred interface block
       shall be the name of a deferred argument of the scoping unit in
       which the <interface-body> appears."
}

[336:2] In last sentence of p2, replace "with neither Abstract nor" with "without ABSTRACT, DEF        ERRED, or" such that the sentence now reads:

    "An interface block without ABSTRACT, DEFERRED, or a generic
     specification is a specific interface block."

> [336:2+] Insert a new paragraph:

Reviewer: ```suggestion
[336:3+] Insert a new paragraph:
```

      SUBROUTINE SUB_SAME_RANK(X)
        INTEGER, RANK(N) :: X
      END SUBROUTINE
    END TEMPLATE"

[344:19+] In section 15.5.1 Syntax of a procedure reference, append
          two new lines to R1522 procedure-designator:

      "  <<or>> <deferred-proc>
         <<or>> <inline-instantiation>"

  such that the rule now reads:

   "R1522 <procedure-designator> <<is>> <procedure-name>
                                 <<or>> <proc-component-ref>
                                 <<or>> <data-ref> % binding-name
                                 <<or>> <deferred-proc>
                                 <<or>> <inline-instantiation>"

[355:5+] Append 2 lines to R1525 actual-arg:

      "<<or>> <deferred-proc>
       <<or>> <inline-instantiation>

  Extend R1524 <actual-arg> with <deferred-proc> and

>   <inline-instantiation> such that it reads:

Reviewer: ```suggestion
[345:5+] Extend R1524 <actual-arg> with <deferred-proc> and
  <inline-instantiation> such that it reads:
```

          shall not be of a deferred type.  It shall have the same
          shape, type, and type parameter values, in corres ponding
          references. It shall not be a coindexed object. It is an
          INTENT (INOUT) argument.  If A is scalar, the computed value
          is the result of the reduction operation of applying OPERA
          TION to the values of A in all corresponding references. If
          A is an array, each element of the computed value is equal
          to the result of the reduction operation of applying
          OPERATION to corresponding elements of A in all
          corresponding references.  The computed value is assigned to
          A if no error condition occurs, and either RESULT_IMAGE is
          absent, or the executing image is the one identified by
          RESULT_IMAGE. Otherwise, A becomes undefined."

{UTI:  Need something in 19.5.1 association}

Annex A:
--------

* A.2 Processor Dependencies

[597:43] Replace final period "." with semicolon ";".

[597:43+] Append a new bullet in the list of processor dependencies

    * the names of the dummy arguments of a deferred procedure without

>       an explicit specification (tt.8.2.4).

Reviewer: How confident are we that this is the correct section number?
